engine/src/main/java/org/hibernate/validator/internal/util/TypeHelper.java

```diff
@@ -324,24 +326,32 @@ private static Type getResolvedSuperclass(Type type) {
 		return validatorsTypes;
 	}
 
-	public static Type extractType(Class<? extends ConstraintValidator<?, ?>> validator) {
-		Map<Type, Type> resolvedTypes = newHashMap();
+	public static Type extractConstraintValidatorValidatedType(Class<? extends ConstraintValidator<?, ?>> validator) {
+		return extractConstraintValidatorTypeArgumentType( validator, VALIDATOR_TYPE_INDEX );
+	}
+
+	public static Type extractConstraintValidatorConstraintType(Class<? extends ConstraintValidator<?, ?>> validator) {
+		return extractConstraintValidatorTypeArgumentType( validator, CONSTRAINT_TYPE_INDEX );
+	}
+
+	public static Type extractConstraintValidatorTypeArgumentType(Class<? extends ConstraintValidator<?, ?>> validator, int typeArgumentIndex) {
+		Map<Type, Type> resolvedTypes = new HashMap<>();
 		Type constraintValidatorType = resolveTypes( resolvedTypes, validator );
 
 		//we now have all bind from a type to its resolution at one level
-		Type validatorType = ( (ParameterizedType) constraintValidatorType ).getActualTypeArguments()[VALIDATOR_TYPE_INDEX];
-		if ( validatorType == null ) {
+		Type constraintType = ( (ParameterizedType) constraintValidatorType ).getActualTypeArguments()[typeArgumentIndex];
```

Comment:
Shouldn't it be just `type` here? Or perhaps `extractedType`. After all, it can be the constraint type _or_ the validator type.

Reply:
Yup, it's coming from a late minute refactoring, will fix.